src/PathFinding.h

```diff
@@ -0,0 +1,91 @@
+#include <vector>
+#include <map>
+#include <functional>
+#include <algorithm>
+#include <limits>
+
+#ifndef PATH_FINDING_H
+#define PATH_FINDING_H
+
+template<typename Node, typename Key>
+std::vector<Node*> find_shortest_path(
+    Node& start,
+    std::function<Key (Node&)> get_key,
+    std::function<std::vector<Node*> (Node&)> get_neighbors,
```

Comment:
C++ trained me to be alarmed when seeing so many `std::vectors` being created, although you would probably do this in other programming languages.

Reply:
To be honest, I felt pretty weird spamming `std:vector` but didn't know how to do this more idiomatically